### PR TITLE
Start writing migration guide for Gradle plugin apply

### PR DESCRIPTION
**Pull request by bartekpacia**
This PR adds a guide explaining how to migrate from _imperative apply_ of Flutter's Gradle plugins to _declarative apply_.

Resolves https://github.com/flutter/flutter/issues/135392

## Presubmit checklist

- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

## Preview

[here](https://flutter-docs-prod--pr9857-migration-guide-gradle-plugi-vyllekcb.web.app/release/breaking-changes/flutter-gradle-plugin-apply)

### CODE REVIEW

**Comment by bartekpacia** [comment]
waiting for https://github.com/flutter/flutter/pull/139006 to be merged before proceeding further

**Comment by bartekpacia** [comment]
TODO (probably discuss): Add `logger.warn("This is deprecated - don't use)"` (but ofc prettier) to `flutter.gradle` and `app_plugin_loader.gradle`.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (johnmccutchan) [inline src/release/breaking-changes/flutter-gradle-plugin-apply.md:21]: Should this be an actual link to the Gradle docs?

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (johnmccutchan) [inline src/release/breaking-changes/flutter-gradle-plugin-apply.md:32]: Be more explicit:

Replace the contents of `android/settings.gradle` with:

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (johnmccutchan) [inline src/release/breaking-changes/flutter-gradle-plugin-apply.md:65]: be more explicit, Replace the contents of `android/build.gradel` with:

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (johnmccutchan) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:32]: but also, what if people have made additions / changes to this file? Perhaps the "replace" should be more specific.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (johnmccutchan) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:65]: but also, what if people have made additions / changes to this file? Perhaps the "replace" should be more specific.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:16]: Agreed.

[Gradle itself refers](https://docs.gradle.org/current/userguide/plugins.html#sec:plugins_block) to `plugins {}` as "`plugins {}` block", so my proposal is:
- imperative _apply script method_ is the old way of applying Gradle plugins
- _declarative plugins {} block_ is the new way of applying Gradle plugins

I don't want to bikeshed over this too much so I'll let you make the final call and I'm fine with whatever.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:21]: It is. At the bottom of the page there's:

```
[plugins block]: https://docs.gradle.org/current/userguide/plugins.html#plugins_dsl_limitations
```

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: agreed

done in 13507699bb09b0438555f56223e8a2564def76f0

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (reidbaker) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:16]: I like the delta between the 2 you listed. I agree on wanting to avoid bikeshedding. 

* imperative apply script method is the old way of applying Gradle plugins
* declarative plugins {} block is the new way of applying Gradle plugins

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (Gustl22) [inline src/release/breaking-changes/flutter-gradle-plugin-apply.md:189]:
```suggestion
For a detailed overview, see [issue #135392][]. You should consider using the
```

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: thanks, I mostly agree, done in 13507699bb09b0438555f56223e8a2564def76f0. I went with:

> Projects created with versions of Flutter prior to 3.16 need to be migrated manually.

because I'm not aware of any work happening on `flutter migrate`, which is the only potential solution to automate this migration.

Also I understand your good intentions with this:

> the build migration (link to steps) should take one engineer less than a day.

but we don't have data to back that up (except personal experience). I've seen big Flutter projects that have a lot of custom Gradle conifg and I just wouldn't be comfortable with saying this (haha, I'm referring to personal experience now, but you get me). I think it might sometimes take more than 1 day, esp. if the engineer is very new to Gradle. I'd not include it.

I included this:

> Applying Gradle plugins using the `plugins {}` executes the same code as before
and should produce equivalent app binaries.

in 3308f5aaee0b2a4bdb3ce62a73627721ce53aedd

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:59]: Sure – would this:
```gradle
id "com.android.application" version "{agpVersion}" apply false
id "org.jetbrains.kotlin.android" version "{kotlinVersion}" apply false
```
be ok?

PS I was also thinking about including the newest possible versions of AGP and Kotlin here, because I think many people are not aware these things actually get updates and are stuck on AGP v4, haha.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:12]: Do you mean [`flutter.gradle`](https://github.com/flutter/flutter/blob/flutter-3.13-candidate.0/packages/flutter_tools/gradle/flutter.gradle)? Or maybe the `android/app/build.gradle` of a real example app? If so, which one?

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: done in a7b28e38ff8998e0d75b0244dddce8c2361039cb

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:26]: > Migrating the app ecosystem to use this approach will eventually be required, see flutter deprecation timeline (with link), It also is a prerequisite for flutter migrating to kotlin and off of groovy which we believe will lead to greater build stability.

Should this be included in this doc?

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:172]: tbh I'm not sure if it's required. But it's a good practice and I think it's worth demonstrating it on a common example.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (Gustl22) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:59]: Only a quick thought, don't know if it's practical and can be ignored: can't we make a test, if the docs code is equivalent to the template, so it would replicate the up to date version of this one. But it does of course not work with more complicated code snippets like below.

I always find it helpful to know what is the Flutter supported version or at least, have a link to the updated template file.

Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:59]: That'd be great – but I think it's out of scope of this PR (it deserves a separate issue imho)

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:122]: done in 93a4b2abdfa1230d25a97d842c757b109f863531!

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: done!

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: Agreed. I rephreased this in 820544683aed2191c91f7cb25c2b4fcd8683bd7e

But making changes to `settings.gradle` is rare (at least according to my experience as Flutter developer), changing `build.gradle` and `app/build.gradle` is done much more often.

**Review by sfshaza2 — APPROVED** [review]
lgtm
@reidbaker @bartekpacia, it _appears_ that you have come to agreement on this PR. I am approving it, with that understanding. Is it ok if we merge this? By the way, thanks so much for your conscientious help!!

**Comment by bartekpacia** [comment]
Thank you @sfshaza2! As of now I'm waiting for @reidbaker to review my changes to his comments.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (reidbaker) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:59]: Please remove this version before merging. 
"{{agpVersion}} works for me so does <AGP-Version> to mirror how you document <app-src>

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (reidbaker) [inline src/release/breaking-changes/flutter-gradle-plugin-apply.md:136]: Before the example I think we should have a section that gives a user a way to validate they migrated successfully. 

Perhaps something like 
```
### Validation 

Execute  `flutter run apk` and confirm your app builds and launches. 

## Examples 
```

**Review by reidbaker — APPROVED** [review]
Approved, assuming the version for AGP in the example is removed.

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md:147]: done in b7f022f93e7798f3b5eec6b69027372d56ac4762

File: src/release/breaking-changes/flutter-gradle-plugin-apply.md
Comment (bartekpacia) [reply src/release/breaking-changes/flutter-gradle-plugin-apply.md]: done in 35e58151a062f23a009220fab778b99998d3dadc

**Comment by sfshaza2** [comment]
Congrats on your first PR, @bartekpacia!

**Comment by dumpster-of-things** [comment]
Inconsistent syntax used in the documentation
( https://docs.flutter.dev/release/breaking-changes/flutter-gradle-plugin-apply )

"Most changes have to be made in the <app-src>/android/app/build.gradle. " ... "Now apply the plugins again, but this time using the Plugin DSL syntax. At the very top of your file, add:"
```
+plugins {
+    id "com.android.application"
+    id "kotlin-android"
+    id "dev.flutter.flutter-gradle-plugin"
+}
```

* Shortly thereafter, the same block of code is referenced (seemingly) inconsistently:
"and the following lines to <app-src>/android/app/build.gradle:"
```
plugins {
     id "com.android.application"
     id "dev.flutter.flutter-gradle-plugin"
     id "org.jetbrains.kotlin.android"
+    id "com.google.gms.google-services"
+    id "com.google.firebase.crashlytics"
 }
```

**Comment by bartekpacia** [comment]
In reply to [this comment](https://github.com/flutter/website/pull/9857#issuecomment-1921097968):

That's not true. You cannot use "kotlin-android" when specifying plugin coordinates in the `plugins {}` block in `settings.gradle`. Try and see that it doesn't work.

**Pull request merged**
